server/src/main/java/org/elasticsearch/action/admin/cluster/stats/TransportClusterStatsAction.java

```diff
@@ -315,4 +377,119 @@ protected boolean isFresh(Long currentKey, Long newKey) {
             return newKey <= currentKey;
         }
     }
+
+    private static boolean doRemotes(ClusterStatsRequest request) {
+        return CCS_TELEMETRY_FEATURE_FLAG.isEnabled() && request.doRemotes();
+    }
+
+    private class RemoteStatsFanout extends CancellableFanOut<String, RemoteClusterStatsResponse, Map<String, RemoteClusterStats>> {
+        private final ClusterStatsRequest request;
+        private final Map<String, RemoteClusterStatsResponse> responses = new ConcurrentHashMap<>();
+        private final Executor requestExecutor;
+        private final Task task;
+        private final TaskId taskId;
+
+        RemoteStatsFanout(Task task, ClusterStatsRequest request, Executor requestExecutor) {
+            this.task = task;
+            this.request = request;
+            this.requestExecutor = requestExecutor;
+            if (task instanceof CancellableTask cancellableTask) {
+                cancellableTask.addListener(responses::clear);
+            }
+            this.taskId = new TaskId(clusterService.getNodeName(), task.getId());
+        }
+
+        @Override
+        protected void sendItemRequest(String clusterAlias, ActionListener<RemoteClusterStatsResponse> listener) {
+            var remoteRequest = new RemoteClusterStatsRequest(request.nodesIds());
+            var remoteClusterClient = remoteClusterService.getRemoteClusterClient(
+                clusterAlias,
+                requestExecutor,
+                RemoteClusterService.DisconnectedStrategy.RECONNECT_IF_DISCONNECTED
+            );
+            remoteRequest.setParentTask(taskId);
+            remoteClusterClient.execute(TransportRemoteClusterStatsAction.REMOTE_TYPE, remoteRequest, listener);
+        }
+
+        @Override
+        protected void onItemResponse(String clusterAlias, RemoteClusterStatsResponse response) {
+            if (isCancelled() == false) {
+                responses.put(clusterAlias, response);
+            }
+        }
+
+        @Override
+        protected void onItemFailure(String clusterAlias, Exception e) {
+            logger.warn("Failed to get remote cluster stats for [{}]: {}", clusterAlias, e);
+        }
+
+        private boolean isCancelled() {
+            return task instanceof CancellableTask cancellableTask && cancellableTask.isCancelled();
+        }
+
+        @Override
+        protected Map<String, RemoteClusterStats> onCompletion() {
+            if (isCancelled()) {
+                return Map.of();
+            }
+
+            Map<String, RemoteClusterStats> remoteClustersStats = new HashMap<>();
+
+            for (String clusterAlias : remoteClusterService.getRegisteredRemoteClusterNames()) {
```

Comment:
This might return a different collection of remotes from the one over which we originally iterated, if the collection changed concurrently. Could we instead iterate over this just once, capturing the info about the remote connection when sending the request or receiving the response?

Reply:
Interesting scenario, didn't think about this possibility. I'm not sure if it's easy to eliminate second iteration altogether, but it's certainly not hard to ensure both work on the same list, I'll ensure that.